app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusHandlerThread.java

```diff
@@ -0,0 +1,446 @@
+package nodomain.freeyourgadget.gadgetbridge.service.devices.hplus;
+
+/*
+* @author João Paulo Barraca &lt;jpbarraca@gmail.com&gt;
+*/
+
+
+import android.content.Context;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+
+import nodomain.freeyourgadget.gadgetbridge.GBApplication;
+import nodomain.freeyourgadget.gadgetbridge.GBException;
+import nodomain.freeyourgadget.gadgetbridge.database.DBHandler;
+import nodomain.freeyourgadget.gadgetbridge.database.DBHelper;
+import nodomain.freeyourgadget.gadgetbridge.devices.hplus.HPlusConstants;
+import nodomain.freeyourgadget.gadgetbridge.devices.hplus.HPlusHealthSampleProvider;
+import nodomain.freeyourgadget.gadgetbridge.entities.DaoSession;
+import nodomain.freeyourgadget.gadgetbridge.entities.HPlusHealthActivityOverlay;
+import nodomain.freeyourgadget.gadgetbridge.entities.HPlusHealthActivityOverlayDao;
+import nodomain.freeyourgadget.gadgetbridge.entities.HPlusHealthActivitySample;
+import nodomain.freeyourgadget.gadgetbridge.impl.GBDevice;
+import nodomain.freeyourgadget.gadgetbridge.model.ActivityKind;
+import nodomain.freeyourgadget.gadgetbridge.model.ActivitySample;
+import nodomain.freeyourgadget.gadgetbridge.service.btle.TransactionBuilder;
+import nodomain.freeyourgadget.gadgetbridge.service.serial.GBDeviceIoThread;
+
+
+class HPlusHandlerThread extends GBDeviceIoThread {
```

Comment:
You're totally right, we're lacking such a thing. I'd love to have some generic support for such things though. We definitely need it for automatic activity sync with other devices as well.

I had imagined to implement something a bit like anacron though. Interested parties configure     a time/interval and the anacron service invokes all clients accordingly. It would use `AlarmManager` to wake up when necessary, and it would attempt to batch invocations in order to minimize wakeups.

It would also handle failures (e.g. when a device is not connected) and attempt to repeat the invocation.

Haven't gotten around to implementing that, though.